libcontainerd/remote/client.go

```diff
@@ -511,26 +511,44 @@ func (c *client) Status(ctx context.Context, containerID string) (containerd.Pro
 	return s.Status, nil
 }
 
+func (c *client) getCheckpointOptions(id string, exit bool) containerd.CheckpointTaskOpts {
+	return func(r *containerd.CheckpointTaskInfo) error {
+		if !exit {
+			// shortcut no-op for now since exit is the only thing we set
+			return nil
+		}
+
+		c.v2runcoptionsMu.Lock()
+		_, isV2 := c.v2runcoptions[id]
+		c.v2runcoptionsMu.Unlock()
+
+		if isV2 {
```

Comment:
Can we maybe do something like this instead?
```go
if r.Options == nil {
    if isV2 {
        r.Options = &v2runcoptions.CheckpointOptions{}
    } else {
        r.Options = &runctypes.CheckpointOptions{}
    }
}
r.Options.Exit = exit
```

This seems slightly more readable as well as potentially extensible.

Maybe remove the `if !exit` shortcut, too, unless it saves us a lot of resources (which I doubt).

Reply:
Your code doesn't work because of the different types.
`Options` is an `interface{}` that we need to assert against... but certainly we can do something like

```go
if r.Options == nil {
    if isV2 {
        r.Options = &v2runcoptions.CheckpointOptions{}
    } else {
        r.Options = &runctypes.CheckpointOptions{}
    }
}

switch o := r.Options.(typ) {
// assertions
}
```

Reply:
I wanted to avoid assertions where it's not needed as the old code did... but in the grand scheme of things it's not saving us anything noticeable.

Reply:
> Your code doesn't work because of the different types.

right :man_facepalming: